Create pull request with quick pick

rebornix opened the pull request:
This PR reflects the recording shared in https://github.com/Microsoft/vscode-pull-request-github/issues/620#issuecomment-433568713. It derived from https://github.com/Microsoft/vscode-pull-request-github/commit/534dd0015cbd0a99c4c09844f4d8aa1161b4fb9b in @queerviolet 's pull request #554 and the code change is minimal 

* Detect local remotes and their default targets, basically that's the logic in #554 
* Users only choose target remote and branch
* We show a progress bar while creating Pull Request
* Switch to Review Mode immediately and open the description view for that PR. We already merged https://github.com/Microsoft/vscode-pull-request-github/pull/643 so users can update the description of the pull request with ease.

https://github.com/Microsoft/vscode-pull-request-github/pull/554 shows all the possibilities of creating pull request: quick pick (modal), webview and `pullrequest.md`, and it inroduces Preact and Redux, which simplies the webview rendering code and improves the testability significantly. For short, we love both the UX and engineering improvement in #554. Instead of going back and forth there, let's branch out and use this pr to bring in the *first iteration* workflow.

@queerviolet I'd love to hear your ideas on how we can bring remote/branch logic from https://github.com/Microsoft/vscode-pull-request-github/pull/554/files#diff-16d81de09f288c2fbde3485db2c8003f into this pr. We can discuss this on Monday's syncup.

rebornix commented:
@queerviolet if you see anything missing from #554, feel free to push directly to this branch or leave comments, both ways work.

RMacfarlane commented on src/common/git.ts:
```diff
@@ -0,0 +1,62 @@
+ * @param {string?} localBranch name of local branch to push (default: current branch)
+ * @param {string?} remoteBranch name of remote branch (default: name of local branch)
+ */
+export async function push(
```
we should be able to use the git extension API for this, it was added recently. would require bumping engine to 1.29

rebornix replied:
yes we should do this.

RMacfarlane replied:
we can remove this file, not used anywhere

RMacfarlane commented on src/github/pullRequestManager.ts:
```diff
@@ -397,6 +441,98 @@ export class PullRequestManager implements IPullRequestManager {
+		}
+
+		// Create PR
+		let { data } = await repo.octokit.pullRequests.create(params);
```
we should try-catch this here or in the calling code

RMacfarlane commented on src/github/githubRepository.ts:
```diff
@@ -36,15 +37,20 @@ export class GitHubRepository implements IGitHubRepository {
-		try {
-			const { octokit, remote } = await this.ensure();
-			const { data } = await octokit.repos.get({
+	async getMetadata(): Promise<any> {
```
in master there's a change to bump the version of the `octokit/rest` library, which adds return type information, so it should be easy to create an interface that extends that type to have `currentUser`

rebornix replied:
agree and ideally we can have `currentUser` in `@octokit/rest` typings as well.

RMacfarlane commented on src/view/reviewManager.ts:
```diff
@@ -1025,6 +1028,51 @@ export class ReviewManager implements vscode.DecorationProvider {
+			let selectedRemote = selected.remote;
+			const base: any = await this._prManager.getMetadata(selectedRemote.remoteName);
+			let targets = `${base.owner.login}:${base.default_branch}`;
+			vscode.window.showQuickPick([targets]).then(target => {
```
```ts
if (!target) {
	return
}
```

queerviolet replied:
It doesn't seem like there's a reason to nest `.then`s here, rather than chaining them or using `await`.

RMacfarlane commented on src/view/reviewManager.ts:
```diff
@@ -1025,6 +1028,51 @@ export class ReviewManager implements vscode.DecorationProvider {
+		let picks: vscode.QuickPickItem[] = potentialTargetRemotes.map(remote => new RemoteQuickPickItem(remote));
+		vscode.window.showQuickPick(picks, {
+			ignoreFocusOut: true,
+			placeHolder: 'Choose a remote which you want to send pull request to'
```
small typo: `send pull request to` -> `send a pull request to`

RMacfarlane commented on src/view/reviewManager.ts:
```diff
@@ -1025,6 +1028,51 @@ export class ReviewManager implements vscode.DecorationProvider {
+			}
+			let selectedRemote = selected.remote;
+			const base: any = await this._prManager.getMetadata(selectedRemote.remoteName);
+			let targets = `${base.owner.login}:${base.default_branch}`;
```
should we add something like `Choose a base branch` as the placeholder?

RMacfarlane commented on src/view/reviewManager.ts:
```diff
@@ -1025,6 +1028,55 @@ export class ReviewManager implements vscode.DecorationProvider {
 	}
 
+	public async createPullRequest(): Promise<void> {
+		const branchName = this._repository.state.HEAD.name;
```
right now if you try to create a PR from local branch that hasn't been pushed to the remote, a notification with just "Validation Failed" is shown. should we push the branch in that case?

I think there may also be a problem when the local branch is named differently than the remote branch

rebornix replied:
for the first iteration, we can show a warning for that case. We didn't decide yet how we are going to handle the `push` workflow.

queerviolet commented on src/view/reviewManager.ts:
```diff
@@ -1025,6 +1028,55 @@ export class ReviewManager implements vscode.DecorationProvider {
 
+	public async createPullRequest(): Promise<void> {
+		const branchName = this._repository.state.HEAD.name;
+		let potentialTargetRemotes = this._prManager.getGitHubRemotes();
```
We could call `getPullRequestDefaults` here to get the default values for the quickpicks. That'll handle guessing reasonable initial values.

rebornix commented:
Addressed comments except https://github.com/Microsoft/vscode-pull-request-github/pull/661#discussion_r230855623 (we can update this after we discuss how to handle `push` like in #554) and https://github.com/Microsoft/vscode-pull-request-github/pull/661#discussion_r230868193 (which can probably be updated in upstream).

RMacfarlane commented on src/github/pullRequestManager.ts:
```diff
@@ -748,3 +894,36 @@ export function getEventType(text: string) {
+
+const ownedByMe: Predicate<GitHubRepository> = repo => {
+	const { currentUser=null } = repo.octokit as any;
+	return repo.remote.owner === currentUser.login;
```
let's check that `currentUser` exists before dereferencing `login`

rebornix replied:
done

RMacfarlane commented on src/github/pullRequestManager.ts:
```diff
@@ -748,3 +894,36 @@ export function getEventType(text: string) {
+	({remote: {owner, repositoryName}}) => byOwner === owner && repo === repositoryName;
+
+const titleAndBodyFrom = (message: string): {title: string, body: string} => {
+	const idxLineBreak = message.indexOf('\n');
```
would message ever contain `\r\n` line endings?

RMacfarlane commented on src/view/reviewManager.ts:
```diff
@@ -1024,6 +1027,65 @@ export class ReviewManager implements vscode.DecorationProvider {
 
+	public async createPullRequest(): Promise<void> {
+		const HEAD = this._repository.state.HEAD;
+		if (!HEAD.upstream) {
```
`upstream` seems to always be undefined

rebornix replied:
discussed offline, the local branch needs to be `--set-upstream` to the remote one.

hanif147 commented:
https://www.imdb.com/list/ls095865806/
